GearRent/Networking/APIManager.m

```diff
@@ -0,0 +1,144 @@
+//
+//  APIManager.m
+//  GearRent
+//
+//  Created by Edwin Delgado on 7/20/22.
+//
+
+#import "APIManager.h"
+#import "Parse/Parse.h"
+#import "GNGeoHash.h"
+#import "CoreLocation/CoreLocation.h"
+#import "Edge.h"
+
+@implementation APIManager
+
++ (void)fetchListingsWithCoordinates:(NSArray *)coordinates completion:(void(^)(NSArray *listings, NSError *error))completion{
+    // Find all the geohashes within the polygon and return nil if no geohashes with precision 7 exist
+    NSMutableSet *geohashesWithinPolygon = [[NSMutableSet alloc] init];
+    for(int i = 1; i <= 7; i ++){
+        geohashesWithinPolygon = [self findAllGeohashesWithinPolygon:coordinates precision:i];
+        if([geohashesWithinPolygon count] > 0){
+            break;
+        }
+    }
+    if([geohashesWithinPolygon count] == 0){
+        NSError *error = [NSError errorWithDomain:@"someDomain" code:404 userInfo:@{@"Error reason": @"Polygon too small"}];
+        completion(nil, error);
+    }
+    [PFCloud callFunctionInBackground:@"getListingsWithGeohashes"
+                       withParameters:@{@"geohashes": [geohashesWithinPolygon allObjects]}
+                                block:^(NSArray *listings, NSError *error) {
+      if (!error) {
+          NSLog(@"%@", listings); // complete with listings after
+      }
+    }];
+}
+
++ (NSMutableSet *)findAllGeohashesWithinPolygon:(NSArray *)coordinates precision:(int)precision {
+    NSMutableSet *uncheckedGeohashes = [[NSMutableSet alloc] init];
+    NSMutableSet *insideGeohashes = [[NSMutableSet alloc] init];
+    NSMutableSet *outsideGeohashes = [[NSMutableSet alloc] init];
+    for(int i = 0; i < coordinates.count; i++){
+        CLLocation *location = (CLLocation *) coordinates[i];
+        GNGeoHash *gh = [GNGeoHash withCharacterPrecision: location.coordinate.latitude andLongitude:location.coordinate.longitude andNumberOfCharacters:precision];
+        [uncheckedGeohashes addObject:gh];
+    }
+    while (uncheckedGeohashes.count > 0) {
+        GNGeoHash *currGH = (GNGeoHash *)[uncheckedGeohashes allObjects][0];
```

Comment:
use `uncheckedGeohashes.firstObject` is safer as it'll return nil when the array is empty while accessing 0 index might crash

Reply:
I'm using sets to keep the constant lookup time